segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/MetadataStore.java

```diff
@@ -0,0 +1,717 @@
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded.
+     */
+    CompletableFuture<Void> updateSegmentInfo(SegmentMetadata segmentMetadata, Duration timeout) {
+        if (segmentMetadata.isMerged()) {
+            return Futures.failedFuture(new StreamSegmentMergedException(segmentMetadata.getName()));
+        } else if (segmentMetadata.isDeleted()) {
+            return Futures.failedFuture(new StreamSegmentNotExistsException(segmentMetadata.getName()));
+        }
+
+        ArrayView toWrite = SegmentInfo.serialize(new SegmentInfo(segmentMetadata.getId(), segmentMetadata.getSnapshot()));
+        return updateSegmentInfo(segmentMetadata.getName(), toWrite, timeout);
+    }
+
+    /**
+     * Updates information about a Segment.
+     *
+     * @param segmentName The Segment name.
+     * @param segmentInfo An {@link ArrayView} representing the serialized form of a {@link SegmentInfo} that will be
+     *                    written to the Metadata Store.
+     * @param timeout     Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded.
+     */
+    protected abstract CompletableFuture<Void> updateSegmentInfo(String segmentName, ArrayView segmentInfo, Duration timeout);
+
+    //endregion
+
+    //region Segment Id Assignment
+
+    /**
+     * Attempts to get an existing SegmentId for the given Segment Name, and then invokes the given Function with the Id.
+     * * If the Segment is already mapped in the Metadata, the existing Id is used.
+     * * Otherwise if the Segment had previously been assigned an id (and saved in the Metadata Store), that Id will be
+     * reused.
+     * * Otherwise, it atomically assigns a new Id and stores it in the Metadata and DurableLog.
+     *
+     * If multiple requests for assignment arrive for the same Segment in parallel (or while an assignment is in progress),
+     * they will be queued up in the order received and will be invoked in the same order after assignment.
+     *
+     * @param segmentName The Segment Name.
+     * @param timeout     The timeout for the operation.
+     * @param thenCompose A Function that consumes a SegmentId and returns a CompletableFuture that will indicate
+     *                    when the consumption of that SegmentId is complete. This Function will be invoked
+     *                    synchronously if the SegmentId is already mapped, or async, otherwise, after assignment.
+     * @param <T>         Type of the return value.
+     * @return A CompletableFuture that, when completed normally, will contain the result of the given Function (thenCompose)
+     * applied to the assigned/retrieved SegmentId. If failed, this will contain the exception that caused the failure.
+     */
+    <T> CompletableFuture<T> getOrAssignSegmentId(String segmentName, Duration timeout, @NonNull Function<Long, CompletableFuture<T>> thenCompose) {
```

Comment:
Because of `thenCompose`. For multiple concurrent calls on the same segment, we need a mechanism which will invoke the callbacks in the same order in which they were executed by the external API. Using `CompletableFuture.thenCompose` does not guarantee that, hence the complicated mechanism in this class which keeps a queue of callbacks. 

The return of this is whatever `thenCompose` returns, since that's what the external caller wants.